testing/citest/tests/kube_v2_smoke_test.py

```diff
@@ -0,0 +1,273 @@
+                    'name': name,
+                    'namespace': self.TEST_NAMESPACE,
+                    'labels': {
+                        'app': self.TEST_APP,
+                    }
+                },
+                'spec': {
+                    'replicas': 1,
+                    'selector': {
+                        'matchLabels': {
+                            'app': self.TEST_APP,
+                        }
+                    },
+                    'template': {
+                        'metadata': {
+                            'labels': {
+                                'app': self.TEST_APP,
+                            }
+                        },
+                        'spec': {
+                            'containers': [{
+                                'name': 'primary',
+                                'image': self.__desired_image_id,
+                            }]
+                        }
+                    }
+                }
+            }]
+        }],
+        description='Deploy manifest',
+        application=self.TEST_APP)
+
+    builder = kube.KubeContractBuilder(self.kube_v2_observer)
+    (builder.new_clause_builder('Deployment created',
+                                retryable_for_secs=15)
+     .get_resources(
+         'deploy',
+         extra_args=[name, '--namespace', self.TEST_NAMESPACE])
+     .EXPECT(ov_factory.value_list_contains(jp.DICT_MATCHES(
+         { 'spec': jp.DICT_MATCHES({ 'replicas': jp.NUM_EQ(1) }) }
+     ))))
+
+    return st.OperationContract(
+        self.new_post_operation(
+            title='deploy_manifest', data=payload, path='tasks'),
+        contract=builder.build())
+
+  def delete_manifest(self):
+    """Creates OperationContract for deleteManifest
+
+    To verify the operation, we just check that the Kubernetes deployment
+    is no longer visible (or is in the process of terminating).
+    """
+    bindings = self.bindings
+    name = self.TEST_APP + '-deployment'
+    payload = self.agent.make_json_payload_from_kwargs(
+        job=[{
+            'cloudProvider': 'kubernetes',
+            'type': 'deleteManifest',
+            'account': bindings['SPINNAKER_KUBERNETES_V2_ACCOUNT'],
+            'user': '[anonymous]',
+            'kinds': [ 'deployment' ],
+            'location': self.TEST_NAMESPACE,
+            'options': { },
+            'labelSelectors': {
+                'selectors': [{
+                    'kind': 'EQUALS',
+                    'key': 'app',
+                    'values': [ self.TEST_APP ]
+                }]
+            }
+        }],
+        application=self.TEST_APP,
+        description='Destroy Manifest')
+
+    builder = kube.KubeContractBuilder(self.kube_v2_observer)
+    (builder.new_clause_builder('Replica Set Removed')
+     .get_resources(
+         'deployment',
+         extra_args=[name, '--namespace', self.TEST_NAMESPACE])
+     .EXPECT(self.__not_found_observation_predicate()))
```

Comment:
I thought before you had a size of 0 here (or not found).
The two are different. I think you want both.

The difference is that you were using 0 size as confirmation that eventually the right thing would happen, willing to trust kubernetes from that point. Here you will actually wait for it to finish deleting (which might have flaky timeouts). It's a bit more rigorous but can take longer and doesnt give much value since we're willing to trust k8s. However other tests that run after this which want the delete to be finished can be assured that their preconditions are met and avoid retries on those operations.

As a rule of thumb (not necessarily right, but at this time I think is fine) I settle for the minimum conditions to give confidence that spinnaker is ok. Thus permit the 0 size. You still need the 404 because you might not see the 0 size transition and 404 is the final steady state. The 0 size is just an acknowledgement that eventually you can expect a 404, which is good enough.

Reply:
The test of size 0 was incorrect, it was only passing because we had stated that a 404 was also OK. The resource never has a desired spec size of 0 after a delete is called.

Reply:
To elaborate, in V1, where your example came from we did hit size 0 at some point during the delete process -- this doesn't happen for the deployment object.

Reply:
I did, but I removed it since it was copied from the v1 test and not appropriate here. The size of 0 will never be reflected in that call, so it's not testing for anything valid. Kubernetes will not render that resource with a desired size of 0, so there is no transition we can even catch in some cases.